File: tests/automatic/worker/worker_api_python/worker_api.py
from __future__ import print_function

import asapo_worker
import json
import sys

def assert_noterr(err, name):
    if err != None:
        print (name + ' err: ', err)
        sys.exit(1)

def assert_err(err, name):
    if err == None:
        print (name + ' err: ', err)
        sys.exit(1)


def assert_metaname(meta,compare,name):
    if meta['name'] != compare:
        print ("error at "+name)
        print ('meta: ', json.dumps(meta, indent=4, sort_keys=True))
        sys.exit(1)

def assert_usermetadata(meta,name):
    if meta['meta']['test'] != 10:
        print ('meta: ', json.dumps(meta, indent=4, sort_keys=True))
        print ("error at "+name)
        print ('meta: ', json.dumps(meta, indent=4, sort_keys=True))
        sys.exit(1)


def assert_eq(val,expected,name):
    if val != expected:
        print ("error at "+name)
        print ('val: ', val,' expected: ',expected)
        sys.exit(1)

def check_single(broker,group_id_new):
    _, meta, err = broker.get_next(group_id_new, meta_only=True)
    assert_noterr(err, "get_next")
    assert_metaname(meta,"1","get next1")
    assert_usermetadata(meta,"get next1")

    data, err = broker.retrieve_data(meta)
    assert_eq(data.tostring().decode("utf-8"),"hello1","retrieve_data data")
    assert_noterr(err, "retrieve_data err")

    _, meta, err = broker.get_next(group_id_new, meta_only=True)
    assert_noterr(err, "get_next2")
    assert_metaname(meta,"2","get next2")
    assert_usermetadata(meta,"get next2")

    _, meta, err = broker.get_last(group_id_new, meta_only=True)
    assert_noterr(err, "get_last1")
    assert_metaname(meta,"5","get last1")
    assert_usermetadata(meta,"get last1")

    _, meta, err = broker.get_next(group_id_new, meta_only=True)
    assert_err(err, "get_next3")

    size,err = broker.get_ndatasets()
    assert_noterr(err, "get_ndatasets")
    assert_eq(size,5,"get_ndatasets")


    err = broker.reset_lastread_marker(group_id_new)
    assert_noterr(err, "reset_lastread_marker")

    _, meta, err = broker.get_next(group_id_new, meta_only=True)
    assert_noterr(err, "get_next4")
    assert_metaname(meta,"1","get next4")
    assert_usermetadata(meta,"get next4")


    _, meta, err = broker.get_by_id(3, group_id_new, meta_only=True)
    assert_noterr(err, "get_by_id")
    assert_metaname(meta,"3","get get_by_id")
    assert_usermetadata(meta,"get get_by_id")

    _, meta, err = broker.get_next(group_id_new, meta_only=True)
    assert_noterr(err, "get_next5")
    assert_metaname(meta,"2","get next5")
    assert_usermetadata(meta,"get next5")


    err = broker.set_lastread_marker(4, group_id_new)
    assert_noterr(err, "set_lastread_marker")

    _, meta, err = broker.get_next(group_id_new, meta_only=True)
    assert_noterr(err, "get_next6")
    assert_metaname(meta,"5","get next6")
    assert_usermetadata(meta,"get next6")


    images,err = broker.query_images("meta.test = 10")
    assert_noterr(err, "query1")
    assert_eq(len(images),5,"size of query answer 1")
    for image in images:
        assert_usermetadata(image,"query_images")


    images,err = broker.query_images("meta.test = 10 AND name='1'")
    assert_eq(len(images),1,"size of query answer 2 ")
    assert_noterr(err, "query2")

    for image in images:
        assert_usermetadata(image,"query_images")

    images,err = broker.query_images("meta.test = 11")
    assert_eq(len(images),0,"size of query answer 3 ")
    assert_noterr(err, "query3")

    images,err = broker.query_images("bla")
    assert_err(err, "wrong query")

def check_dataset(broker,group_id_new):
    id, metas, err = broker.get_next_dataset(group_id_new)
    assert_noterr(err, "get_next_dataset1")
    assert_eq(id,1,"get_next_dataset1")
    assert_metaname(metas[0],"1_1","get nextdataset1 name1")
    assert_metaname(metas[1],"1_2","get nextdataset1 name2")
    assert_usermetadata(metas[0],"get nextdataset1 meta")

    data, err = broker.retrieve_data(metas[0])
    assert_eq(data.tostring().decode("utf-8"),"hello1","retrieve_data from dataset data")
    assert_noterr(err, "retrieve_data from dataset err")


    id, metas, err = broker.get_next_dataset(group_id_new)
    assert_noterr(err, "get_next_dataset2")
    assert_eq(id,2,"get_next_dataset2")
    assert_metaname(metas[0],"2_1","get nextdataset2 name1")

    id, metas, err = broker.get_last_dataset(group_id_new)
    assert_noterr(err, "get_last_dataset1")
    assert_eq(id,10,"get_last_dataset1")
    assert_metaname(metas[2],"10_3","get get_last_dataset1 name3")

    id, metas, err = broker.get_next_dataset(group_id_new)
    assert_eq(id,None,"get_next_dataset3 id")
    assert_eq(id,metas,"get_next_dataset3 metas")
    assert_err(err, "get_next_dataset3 err")

    id, metas, err = broker.get_dataset_by_id(8,group_id_new)
    assert_eq(id,8,"get_dataset_by_id1 id")
    assert_noterr(err, "get_dataset_by_id1 err")
    assert_metaname(metas[2],"8_3","get get_dataset_by_id1 name3")

    id, metas, err = broker.get_next_dataset(group_id_new)
    assert_eq(id,None,"get_next_dataset4 id")


source, path, beamtime, token, mode = sys.argv[1:]

broker, err = asapo_worker.create_server_broker(source,path, beamtime,"",token,1000)

group_id_new, err = broker.generate_group_id()
assert_noterr(err, "generate_group")


if mode == "single":
    check_single(broker,group_id_new)

if mode == "datasets":
    check_dataset(broker,group_id_new)

File: tests/automatic/worker/worker_api/worker_api.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <algorithm>
#include "worker/data_broker.h"
#include "testing.h"

using asapo::M_AssertEq;
using asapo::M_AssertTrue;


struct Args {
    std::string server;
    std::string run_name;
    std::string token;
    std::string datasets;
};

Args GetArgs(int argc, char* argv[]) {
    if (argc != 5) {
        std::cout << "Wrong number of arguments" << std::endl;
        exit(EXIT_FAILURE);
    }
    std::string server{argv[1]};
    std::string source_name{argv[2]};
    std::string token{argv[3]};
    std::string datasets{argv[4]};

    return Args{server, source_name, token, datasets};
}


void TestSingle(const std::unique_ptr<asapo::DataBroker>& broker, const std::string& group_id) {
    asapo::FileInfo fi;
    asapo::Error err;

    err = broker->GetNext(&fi, group_id, nullptr);
    if (err) {
        std::cout << err->Explain() << std::endl;
    }
    M_AssertTrue(err == nullptr, "GetNext no error");
    M_AssertTrue(fi.name == "1", "GetNext filename");
    M_AssertTrue(fi.metadata == "{\"test\":10}", "GetNext metadata");

    asapo::FileData data;
    err = broker->RetrieveData(&fi, &data);
    M_AssertTrue(err == nullptr, "RetrieveData no error");
    M_AssertEq("hello1", std::string(data.get(), data.get() + fi.size));


    err = broker->GetLast(&fi, group_id, nullptr);
    M_AssertTrue(err == nullptr, "GetLast no error");
    M_AssertTrue(fi.name == "10", "GetLast filename");
    M_AssertTrue(fi.metadata == "{\"test\":10}", "GetLast metadata");

    err = broker->GetNext(&fi, group_id, nullptr);
    M_AssertTrue(err != nullptr, "GetNext2 error");

    err = broker->GetLast(&fi, group_id, nullptr);
    M_AssertTrue(err == nullptr, "GetLast2 no error");

    err = broker->GetById(8, &fi, group_id, nullptr);
    M_AssertTrue(err == nullptr, "GetById error");
    M_AssertTrue(fi.name == "8", "GetById filename");

    err = broker->GetNext(&fi, group_id, nullptr);
    M_AssertTrue(err == nullptr, "GetNext3 no error");
    M_AssertTrue(fi.name == "9", "GetNext3 filename");

    auto size = broker->GetNDataSets(&err);
    M_AssertTrue(err == nullptr, "GetNDataSets no error");
    M_AssertTrue(size == 10, "GetNDataSets size");

    err = broker->ResetCounter(group_id);
    M_AssertTrue(err == nullptr, "ResetCounter");

    err = broker->GetNext(&fi, group_id, nullptr);
    M_AssertTrue(err == nullptr, "GetNext4 no error");
    M_AssertTrue(fi.name == "1", "GetNext4 filename");

    auto group_id2 = broker->GenerateNewGroupId(&err);
    err = broker->GetNext(&fi, group_id2, nullptr);
    M_AssertTrue(err == nullptr, "GetNext5 no error");
    M_AssertTrue(fi.name == "1", "GetNext5  filename");

    auto images = broker->QueryImages("meta.test = 10", &err);
    M_AssertTrue(err == nullptr, "query1");
    M_AssertTrue(images.size() == 10, "size of query answer 1");

    images = broker->QueryImages("meta.test = 10 AND name='1'", &err);
    M_AssertTrue(err == nullptr, "query2");
    M_AssertTrue(images.size() == 1, "size of query answer 2");
    M_AssertTrue(fi.name == "1", "GetNext5  filename");


    images = broker->QueryImages("meta.test = 11", &err);
    M_AssertTrue(err == nullptr, "query3");
    M_AssertTrue(images.size() == 0, "size of query answer 3");

    images = broker->QueryImages("meta.test = 18", &err);
    M_AssertTrue(err == nullptr, "query4");
    M_AssertTrue(images.size() == 0, "size of query answer 4");

    images = broker->QueryImages("bla", &err);
    M_AssertTrue(err != nullptr, "query5");
    M_AssertTrue(images.size() == 0, "size of query answer 5");
}


void TestDataset(const std::unique_ptr<asapo::DataBroker>& broker, const std::string& group_id) {
    asapo::FileInfo fi;
    asapo::Error err;

    auto dataset = broker->GetNextDataset(group_id, &err);
    if (err) {
        std::cout << err->Explain() << std::endl;
    }
    M_AssertTrue(err == nullptr, "GetNextDataSet no error");
    M_AssertTrue(dataset.content.size() == 3, "GetNextDataSet size");
    M_AssertTrue(dataset.content[0].name == "1_1", "GetNextDataSet filename");
    M_AssertTrue(dataset.content[2].name == "1_3", "GetNextDataSet filename");
    M_AssertTrue(dataset.content[0].metadata == "{\"test\":10}", "GetNext metadata");

    asapo::FileData data;
    err = broker->RetrieveData(&dataset.content[0], &data);
    M_AssertTrue(err == nullptr, "RetrieveData no error");
    M_AssertEq("hello1", std::string(data.get(), data.get() + dataset.content[0].size));


    dataset = broker->GetLastDataset(group_id, &err);
    M_AssertTrue(err == nullptr, "GetLast no error");
    M_AssertTrue(dataset.content[0].name == "10_1", "GetLastDataset filename");
    M_AssertTrue(dataset.content[0].metadata == "{\"test\":10}", "GetLastDataset metadata");

    dataset = broker->GetNextDataset(group_id, &err);
    M_AssertTrue(err != nullptr, "GetNextDataset2 error");

    dataset = broker->GetLastDataset(group_id, &err);
    M_AssertTrue(err == nullptr, "GetLastDataset2 no error");

    dataset = broker->GetDatasetById(8, group_id, &err);
    M_AssertTrue(err == nullptr, "GetDatasetById error");
    M_AssertTrue(dataset.content[2].name == "8_3", "GetDatasetById filename");

}

void TestAll(const Args& args) {
    asapo::Error err;
    auto broker = asapo::DataBrokerFactory::CreateServerBroker(args.server, ".", args.run_name, args.token, &err);
    broker->SetTimeout(100);
    auto group_id = broker->GenerateNewGroupId(&err);

    if (args.datasets == "single") {
        TestSingle(broker, group_id);
    }
    if (args.datasets == "dataset") {
        TestDataset(broker, group_id);
    }

}



int main(int argc, char* argv[]) {

    auto args = GetArgs(argc, argv);

    TestAll(args);
    return 0;
}
